### PR TITLE
Catheter aq

### PR DESCRIPTION
**Pull request by ibenemerito88**
Extension of openBF for studying the effects of arterial catheters on the blood flow in elastic vessels. The equations have been updated (modification to flux and its gradient, source term, wave speed, Riemann invariants), as well as the conditions for junction, bifucation and anastomosis. 

Windkessel boundary conditions are not supported, reflections are.

Tests for junctions and bifurcations have been included but need to be improved. Test for anastomosis is still missing.

[catheterAQ.pdf](https://github.com/INSIGNEO/openBF/files/4021406/catheterAQ.pdf)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,4 +2,6 @@ test/**/*_results/*
 **/.DS_Store
 cvs_results/*
 docs/build/*
-Manifest.toml
\ No newline at end of file
+Manifest.toml
+*.sh
+*.swp
```
Comment (alemelis) [inline .gitignore]: qui conviene aggiungere anche tutti i `.last` e `.out` così da non averli anche nella PR

File: test/jointcat/script.jl [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+timepoints = range(0, stop=heart.cardiac_T, length=jump)
+current_time = 0.0
+passed_cycles = 0
+counter = 1
+
+dt = openBF.calculateDeltaT(vessels, Ccfl)
+openBF.solveModel(vessels, edges, blood, dt, current_time)
+openBF.updateGhostCells(vessels)
+current_time += dt
+plot(vessels[1].u)
```
Comment (alemelis) [inline test/jointcat/script.jl]: anche questo file forse si può eliminare dalla PR

File: src/bifurcations.jl [hunk 3]
```diff
@@ -46,17 +46,20 @@ end
 Return the Jacobian for bifurcation equations.
 """
 function calculateJacobianBifurcation(v1 :: Vessel, v2 :: Vessel, v3 :: Vessel, U, k)
```
Comment (alemelis) [inline src/bifurcations.jl]: qui non so perché manchi, ma nella signature della funzione si dovrebbe aggiungere la dichiarazione del tipo di `U` e `k`, credo siano due `Array{Float}`

File: src/anastomosis.jl [hunk 4]
```diff
@@ -46,19 +46,20 @@ end
 Return the Jacobian for anastomosis equations.
 """
 function calculateJacobianAnastomosis(v1 :: Vessel, v2 :: Vessel, v3 :: Vessel, U, k)
```
Comment (alemelis) [inline src/anastomosis.jl]: qui, ma anche nelle funzioni successive, per qualche motivo che non ricordo non ho specificato il tipo di `U` e `k`, ma andrebbe fatto. A occhio mi sembrano due `Array{Float,1}`

File: src/anastomosis.jl [hunk 5]
```diff
@@ -46,19 +46,20 @@ end
...
+    @fastmath @inbounds g1 = sqrt(1-v1.Ac/U[4]^4)	# MODIFIED THIS LINE
```
Comment (alemelis) [inline src/anastomosis.jl]: qui puoi usare `U43` e calcolare `g1` come

```
g1 = sqrt(1.0 - v1.Ac / (U43*U[4]))
```

Comment (alemelis) [reply src/anastomosis.jl]: ovviamente `g` va definita dopo `U`

File: src/anastomosis.jl [hunk 6]
```diff
@@ -46,19 +46,20 @@ end
...
+    @fastmath @inbounds J36 = -4.0*k[3]*(v3.Ac/(2*U[6]*g3) + g3)	# MODIFIED THIS LINE
 
-    @fastmath @inbounds J25 =  4.0*k[2]
-
-    @fastmath @inbounds J36 = -4.0*k[3]
-
-    @fastmath @inbounds J41 =  U[4]*U43
-    @fastmath @inbounds J42 =  U[5]*U53
-    @fastmath @inbounds J43 = -U[6]*U63
+    @fastmath @inbounds J41 =  U[4]*U43 - v1.Ac				# MODIFIED THIS LINE
```
Comment (alemelis) [inline src/anastomosis.jl]: e dato che qui si ricalcola `U[4]*U43`, conviene definire una variabile `U44` per questo valore in modo da calcolarla una volta sola

File: src/anastomosis.jl [hunk 7]
```diff
@@ -83,10 +84,17 @@ end
...
-function calculateWstarAnastomosis(U, k)
-    @fastmath @inbounds W1 = U[1] + 4*k[1]*U[4]
-    @fastmath @inbounds W2 = U[2] + 4*k[2]*U[5]
-    @fastmath @inbounds W3 = U[3] - 4*k[3]*U[6]
+function calculateWstarAnastomosis(U, k, vessels :: Array{Vessel,1})
+	# MODIFIED THIS FUNCTION
+    v1 = vessels[1]						# MODIFIED THIS LINE
+    v2 = vessels[2]						# MODIFIED THIS LINE
+    v3 = vessels[3]						# MODIFIED THIS LINE
+    @fastmath @inbounds g1 = sqrt(1-v1.Ac/U[4]^4)		# MODIFIED THIS LINE
```
Comment (alemelis) [inline src/anastomosis.jl]: in generale `*` é più veloce di `^`, quindi puoi riscrivere come `v1.Ac/(U[4]*U[4]*U[4]*U[4])`

in generale se usi direttamente `1.0` invece di `1`, eviti la conversione durante l'esecuzione

File: src/initialise.jl [hunk 8]
```diff
@@ -570,12 +584,16 @@ where gamma_v (`gamma_profile`) is either specified in the vessel definition or
...
-    if haskey(vessel_data, "gamma_profile")
+
+    if haskey(vessel_data,"Rc")                     # MODIFIED THIS LINE
+    return 2*blood.mu*blood.rho_inv*sqrt(15)
+    elseif haskey(vessel_data, "gamma_profile")
         gamma_profile = vessel_data["gamma_profile"]
     else
         gamma_profile = 9
-    end
     return 2*(gamma_profile + 2)*pi*blood.mu*blood.rho_inv
```
Comment (alemelis) [inline src/initialise.jl]: questo dovrebbe andare dopo l'`end`, altrimenti non restituisce quello che vuoi tu nel caso il secondo `if` sia verificato

File: src/solver.jl [hunk 9]
```diff
@@ -32,12 +32,12 @@ end
...
 """
-    waveSpeed(A :: Float64, gamma :: Float64)
+    waveSpeed(A :: Float64, gamma :: Float64, Ac :: Float64)
 
-Compute pulse wave velocity at the given node.
+Compute pulse wave velocity at the given node. It includes the catheter.
 """
-function waveSpeed(A :: Float64, gamma :: Float64)
-    return sqrt(3*gamma*sqrt(A)*0.5)
+function waveSpeed(A :: Float64, gamma :: Float64, Ac :: Float64)
```
Comment (alemelis) [inline src/solver.jl]: cambiando la definizione di questa funzione alcuni test non passano

per semplificare, puoi dare a `Ac` un valore di default (direi `0.0`) che viene usato nel caso in cui la funzione sia chiamata senza quell'argomento

```
function waveSpeed(A::Float64, gamma::Float64, Ac::Float64 = 0.0)
```

Comment (alemelis) [reply src/solver.jl]: oppure cambi i test aggiungendo `Ac` nella chiamata

File: src/anastomosis.jl (hunk 5 above)
Comment (alemelis) [reply src/anastomosis.jl]: intendo dire che le linee da 49 a 54 vanno riorganizzate come 

```
U43 = U[4]*U[4]*U[4]
g1 = sqrt(1.0 - v1.Ac / (U43*U[4]))
U53 = U[5]*U[5]*U[5]
g2 = sqrt(1-v2.Ac/ (U53*U[5]))
U63 = U[6]*U[6]*U[6]
g3 = sqrt(1-v3.Ac/U63*U[6])
```

File: src/initialise.jl [hunk 10]
```diff
@@ -570,12 +584,16 @@ where gamma_v (`gamma_profile`) is either specified in the vessel definition or
...
+    return 2*blood.mu*blood.rho_inv*sqrt(30)
+    elseif haskey(vessel_data, "gamma_profile")
         gamma_profile = vessel_data["gamma_profile"]
     else
         gamma_profile = 9
+    return 2*(gamma_profile + 2)*blood.mu*blood.rho_inv
     end
-    return 2*(gamma_profile + 2)*pi*blood.mu*blood.rho_inv
+
 end
```
Comment (alemelis) [inline src/initialise.jl]: @ibenemerito88  occhio che il return e' sempre dentro l'ultimo `else` (in teoria julia dovrebbe essere abbastanza intelligente da fare il return di `gamma_profile` anche per linea 591...ma non mi affiderei troppo a queste proprieta' esoteriche del linguaggio)


```julia
function computeViscousTerm(vessel_data :: Dict{Any,Any}, blood :: Blood)
    if haskey(vessel_data,"Rc")                     # MODIFIED THIS LINE
        return 2*blood.mu*blood.rho_inv*sqrt(30)
    elseif haskey(vessel_data, "gamma_profile")
        gamma_profile = vessel_data["gamma_profile"]
    else
        gamma_profile = 9
    end
    return 2*(gamma_profile + 2)*blood.mu*blood.rho_inv
end
```

Comment (alemelis) [reply src/initialise.jl]: (sicuri che `pi` non ci vada? tutta la validazione della tesi e' stata fatta in quel modo)
